extensions/amp-viewer-integration/0.1/findtext.js

```diff
@@ -0,0 +1,451 @@
+ * @param {number} start
+ * @param {number} end
+ * @return {?Text}
+ * @param {!Array<!Element>} marked
+ */
+const markSingleTextNode = function(node, start, end, marked) {
+  if (start >= end) {
+    // Do nothing
+    return null;
+  }
+  const parent = node.parentNode;
+  const text = node.wholeText;
+  if (start > 0) {
+    parent.insertBefore(document.createTextNode(
+        text.substring(0, start)), node);
+  }
+  const span = document.createElement('span');
+  span.appendChild(document.createTextNode(text.substring(start, end)));
+  parent.insertBefore(span, node);
+  marked.push(span);
+
+  let endText = null;
+  if (end < text.length) {
+    endText = document.createTextNode(text.substring(end));
+    parent.insertBefore(endText, node);
+  }
+  parent.removeChild(node);
+  return endText;
+};
+
+
+/**
+ * nextTextNode finds the next sibling text node or
+ *   the next text node in the siblings of the parent.
+ * @param {!Text} textNode The node to start to find the next text node.
+ * @return {?Text}
+ */
+const nextTextNode = function(textNode) {
+  // If leaving is true, find the next node from siblings or the parent.
+  // If leaving is false, find the next node from childrens.
+  let leaving = true;
+  /** @type {Node} */
+  let node = textNode;
+  while (true) {
+    if (node == null) {
+      // No more parent.
+      return null;
+    }
+    if (leaving) {
+      const next = node.nextSibling;
+      if (next) {
+        node = next;
+        // visits childrens of node.
+        leaving = false;
+      } else {
+        // back to the parent.
+        node = node.parentNode;
+      }
+      continue;
+    }
+    if (node instanceof Text) {
+      return node;
+    }
+    if (!node.firstChild) {
+      leaving = true;
+    } else {
+      node = node.firstChild;
+    }
+  }
+};
+
+
+/**
+ * A special TextPos object to represent a whitespace injected
+ *   between two block nodes.
+ * @type {!TextPos}
+ */
+const posDomDelimiter = new TextPos(document.createTextNode(' '), 0);
+
+/**
+ * TextScanner visits text nodes under a root node and
+ *   returns charcter positions respectively.
+ */
+export class TextScanner {
+  /**
+   * @param {!Node} root The root node to visit.
+   */
+  constructor(root) {
+    this.internal_ = new TextScannerInternal(root, false);
+    this.next_ = this.internal_.next();
+  }
+
+  /**
+   * next returns the next TextPos.
+   * Returns null when the scanner reaches the end of the text.
+   * @return {?TextPos}
+   */
+  next() {
+    const ret = this.next_;
+    if (ret == null) {
+      return null;
+    }
+    this.next_ = this.internal_.next();
+    if (this.next_ == null && /\s/.test(ret.char)) {
+      // Remove the trailing space.
+      return null;
+    }
+    return ret;
+  }
+}
+
+export class TextScannerInternal {
+  /**
+   * @param {!Node} node The root node to visit.
+   * @param {boolean} needSpace Whether spaces should be output if found.
+   */
+  constructor(node, needSpace) {
```

Comment:
Not entirely sure what `needSpace` does?

Reply:
Simplified whitespace handling in `TextScanner` and removed `needSpace`.
needSpace existed to emulate browsers's whitespace handling.